src/UKHO.S100PermitService.Common/Services/PermitService.cs

```diff
@@ -117,35 +137,93 @@ private void CreatePermitXml(DateTimeOffset issueDate, string dataServerIdentifi
+                        EncryptedKey = EncryptKey(productKey, hardwareId, holding),
+                        Filename = cell.CellCode,
+                        Expiry = holding.ExpiryDate
+                    };
+
+                    if(productsList.Any(x => x.Id == products.Id))
+                    {
+                        productsList.FirstOrDefault(x => x.Id == products.Id).DatasetPermit.Add(dataPermit);
+                    }
+                    else
+                    {
+                        products.DatasetPermit = new List<ProductsProductDatasetPermit> { dataPermit };
+                        productsList.Add(products);
+                    }
+                    products = new();
                 }
-            };
+            }
+            _logger.LogInformation(EventIds.GetProductListCompleted.ToEventId(), "Get Product List from HoldingServiceResponse and ProductKeyService completed for title : {title}", UpnTitle);
             return productsList;
         }
 
-        private static List<ProductKeyServiceRequest> ProductKeyServiceRequest(
+        public bool ValidateSchema(string permitXml, string xsdPath)
+        {
+            var xml = new XmlDocument();
+            xml.LoadXml(permitXml);
+
+            var xmlSchemaSet = new XmlSchemaSet();
+            xmlSchemaSet.Add(null, xsdPath);
+
+            xml.Schemas = xmlSchemaSet;
+
+            var validXml = true;
+            try
+            {
+                xml.Validate((sender, e) =>
+                {
+                    validXml = false;
+                });
+            }
+            catch(XmlSchemaValidationException)
+            {
+                validXml = false;
+                return validXml;
+            }
+            return validXml;
+        }
+
+        private string ReadXsdVersion()
+        {
+            var xsdPath = Path.Combine(_schemaDirectory, "XmlSchema", "Permit_Schema.xsd");
+
+            XmlSchema? schema;
+            using(var reader = XmlReader.Create(xsdPath))
+            {
+                schema = XmlSchema.Read(reader, null);
+            }
+
+            return schema?.Version[..5] ?? null;
+        }
+
+        private List<ProductKeyServiceRequest> ProductKeyServiceRequest(
             IEnumerable<HoldingsServiceResponse> holdingsServiceResponse) =>
             holdingsServiceResponse.SelectMany(x => x.Cells.Select(y => new ProductKeyServiceRequest
             {
                 ProductName = y.CellCode,
                 Edition = y.LatestEditionNumber
             })).ToList();
+
+        private string EncryptKey(IEnumerable<ProductKey> productKey, string hardwareId, HoldingsServiceResponse holding) =>
+            _s100Crypt.CreateEncryptedKey((from str1 in holding.Cells
+                                         join str2 in productKey on str1.CellCode.ToString() equals str2.ProductName
+                                         select str2.DecryptedKey).FirstOrDefault(), hardwareId);
```

Comment:
can split this for better readability. First retrieve the productKey then make a call to _s100Crypt.CreateEncryptedKey.
```suggestion
            var productKey = holding.Cells
    .Join(productKeys,
        cell => cell.CellCode.ToString(),
        key => key.ProductName,
        (cell, key) => key.DecryptedKey)
    .FirstOrDefault();

return _s100Crypt.CreateEncryptedKey(productKey, hardwareId);
```

Reply:
Refactored as suggested